clusterDeployer.py

```diff
@@ -475,6 +472,28 @@ def create_workers(self) -> None:
             node, _ = pair
             node.future.result()
 
+    def _install_worker_with_retry(self, infra_env: str, node: ClusterNode, executor: ThreadPoolExecutor) -> None:
+        def installation_finished(ai: AssistedClientAutomation, node_name: str) -> bool:
+            info = ai.get_ai_host(node_name)
+            return info is not None and info.status in ["error", "added-to-existing-cluster"]
+
+        name = node.config.name
+        for try_count in itertools.count(0):
+            if self._start_node(infra_env, node, False, executor):
+                self._ai.install_ai_host(infra_env, name)
+                common.wait_true(f"installation {name}", installation_finished, ai=self._ai, node_name=name)
+
+                info = self._ai.get_ai_host(name)
+
+                if info is not None and info.status == "added-to-existing-cluster":
+                    logger.info(f"Worker {name} installation finished after {try_count} retries")
+                    break
+
+            logger.warn(f"Worker {name} installation failed, retrying...")
+            node.teardown()
+            self._ai.delete_host(name)
+            node.future.result()
```

Comment:
We don't do nested futures.

Reply:
This is not nested, it is running in different executor.

Reply:
that's what I'm confused about. Why do we use multiple executors (I know you've linked to the docs, but the question is about the code here)? Can't we rewrite the code so that we don't have calls that have to do with futures inside a call running on an executor?

Reply:
The problem is that the virt-install step needs to run in background, we don't have other facility for that other than futures. And because the installation with needs all the installation steps to be done serially we have to use futures for those too otherwise we would install just one by one. I'm open to suggestions, unfortunately I didn't find better solution that would allow us to do installation of workers in parallel.